### PR TITLE
Better diagnostic location for types

### PR DESCRIPTION
**Pull request by pgovind**
I've implemented better diagnostic locations for types implementing preview interfaces. Instead of the diagnostic appearing on the type itself, it now appears on the interface in the interface list. I think this is a good starting point to get feedback first, and if I get sign off here, I can merge this PR and work on improving the diagnostic location for methods and other types with the same approach.

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs [hunk 1]
```diff
@@ -133,6 +133,11 @@ public sealed class DetectPreviewFeatureAnalyzer : DiagnosticAnalyzer
             FieldOrEventIsPreviewTypeRule,
             StaticAbstractIsPreviewFeatureRule);
 
+        protected virtual SyntaxNode? GetPreviewInterfaceNodeForTypeImplementingPreviewInterface(ISymbol typeSymbol, ISymbol previewInterfaceSymbol)
```
Comment (pgovind) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:136]: No VB implementation yet. It's not high priority right now

Comment (pgovind) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:136]: Done.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureTests.Classes.cs [hunk 2]
```diff
@@ -158,5 +158,167 @@ public abstract class AbClass
...
+        [InlineData("struct")]
+        public async Task TestPartialClassOrStructDeclarationEndOfList(string classOrStruct)
+        {
+            var csInput = @$" 
+using System.Runtime.Versioning; using System;
+namespace Preview_Feature_Scratch
+{{
+    public partial {classOrStruct} Zoo";
+
+            csInput += @": NonPreviewInterface, {|#0:IFoo|}
```
[the remark this answers is not in the document]
Comment (pgovind) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureTests.Classes.cs:215]: > do you think it looks better?

Yup.  String interpolation uses `{}` and the diagnostic markers also use `{| |}`. Then I needed to change the code blocks using `{` to `{{` to construct `csInput`, so I decided to just split it at an appropriate place. Is this harder to read though? If it is, then I'll change it in the next PR

**Comment by pgovind** [comment]
I addressed @buyaa-n's comments. Just waiting for sign off from @jmarolf or somebody else from @dotnet/roslyn-analysis

**Review by jmarolf — APPROVED** [review]

**Comment by pgovind** [comment]
Thanks for the reviews @buyaa-n  and @jmarolf!!!

**Pull request merged**